### PR TITLE
feat(orders): add order service and create report excel file

### PR DESCRIPTION
**Pull request by lamvh**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,8 @@
 node_modules/
 dist/
 .env
+server.js
```
Comment (duongdev) [inline .gitignore]: Why do you ignore these files? If they are your local development stuff that other person won't need for development, just don't stage them in the commits.

File: src/index.ts [hunk 2]
```diff
@@ -1,25 +1,47 @@
...
-(async() => {
-    const dir = path.join(__dirname, `/exportData/`)
-
-    if (!fs.existsSync(dir)) {
-        await fs.mkdirSync(dir, {recursive: false})
-    }
-
-     const data = await processData();
-    createXlSXfile({data, dirName: 'src/exportData'})
+const DIR_NAME = process.env.DIR_NAME || "exports";
```
Comment (duongdev) [inline src/index.ts]: Should use `EXPORTS_PATH` instead?

File: src/index.ts [hunk 3]
```diff
@@ -1,25 +1,47 @@
...
+
+(async () => {
+  console.log("\n\n\n");
+  if (!fs.existsSync(DIR_NAME)) {
+    await fs.mkdirSync(DIR_NAME, { recursive: false });
+  }
+
+  const since = moment()
+    .add(-10, "days")
```
Comment (duongdev) [inline src/index.ts]: Why don't use `substract` instead? [Check it here](https://momentjs.com/docs/#/manipulating/subtract/).

File: src/services/index.ts [hunk 4]
```diff
@@ -1,2 +1,3 @@
-export {processData} from './order'
+export {} from './order'
```
Comment (duongdev) [inline src/services/index.ts]: Why import empty?

File: src/services/order.ts [hunk 5]
```diff
@@ -1,86 +1,157 @@
...
-
-const date = moment().subtract(1, 'days').format('DD-MM-YYYY').toString()
-
-// let dirWriteFile = `${__dirname}/exportData/Order-Report-` + date + `.xlsx`
-let dirWriteFile = path.resolve('src/exportData/', 'abc.xlsx')
-if (process.env.NODE_ENV === 'development') {
-    require('dotenv').config()
+import fs, { watchFile } from "fs";
+import path from "path";
+const ORDER_REPORT_HEADERS = [
```
Comment (duongdev) [inline src/services/order.ts]: should be "column" instead of "header"

File: src/services/variant.ts [hunk 6]
```diff
@@ -1,24 +1,42 @@
...
 const request = axios.create({
-    baseURL: 'https://api.storelammoc.vn',
-    timeout: 15e3,
-    headers: {'x-access-token': process.env.TOKEN_KEY},
+  baseURL: "https://api.storelammoc.vn",
+  timeout: 15e3,
+  headers: { "x-access-token": process.env.TOKEN_KEY }
 });
 
+export async function getVariantByIds(ids: string[]) {
```
Comment (duongdev) [inline src/services/variant.ts]: We are getting multiple variants. So this function should be `getVariantsByIds`

File: src/services/order.ts [hunk 7]
```diff
@@ -1,86 +1,157 @@
...
+export const generateXLSXOrderData = async (orders: any[]) => {
+  const variantIds = lodash.flatten(
+    orders.map(order => {
+      return order.items.map((item: any) => item.variant.id);
+    })
+  );
+
+  const variants = await getVariantByIds(variantIds);
+  const variantEntities = lodash.keyBy(variants, "_id");
+  let processedOrders = lodash.flatten(
```
Comment (duongdev) [inline src/services/order.ts]: use `const`. `unshift` does not re-assign the `processedOrders` by a new value.

File: src/services/order.ts [hunk 8]
```diff
@@ -1,86 +1,157 @@
...
+        order.status
+      ]);
+    })
+  );
+  processedOrders.unshift(ORDER_REPORT_HEADERS);
+
+  return processedOrders;
+};
+
+export const fetchOrderDataRecursion = async ({
```
Comment (duongdev) [inline src/services/order.ts]: should be `fetchOrders`

File: src/utils/xlsx.ts [hunk 9]
```diff
@@ -1,11 +1,30 @@
...
-    console.log('Export XLSX Done !')
-}
\ No newline at end of file
+if (process.env.NODE_ENV === "development") {
+  require("dotenv").config();
+}
+
+const DIR_NAME = process.env.DIR_NAME || "exports";
+
+export function createXlSXfile(option: {
```
Comment (duongdev) [inline src/utils/xlsx.ts]: should be `options`

File: src/utils/xlsx.ts [hunk 10]
```diff
@@ -1,11 +1,30 @@
...
+  data: any[][];
+  fileName: string;
+  title: string;
+}) {
+  let data = option.data;
```
Comment (duongdev) [inline src/utils/xlsx.ts]: use `const`

**Review by duongdev — COMMENTED** [review]
Please update this PR with unit test,  Prettier & TSLint configured. Also resolve the change requests below.

File: src/services/order.ts (hunk 7 above)
Comment (vhlam1997) [reply src/services/order.ts]: Before version 2.8.0, the moment#subtract(String, Number) syntax was also supported. It has been deprecated in favor of moment#subtract(Number, String).

File: src/index.ts (hunk 3 above)
Comment (lamvh) [reply src/index.ts]: Before version 2.8.0, the moment#subtract(String, Number) syntax was also supported. It has been deprecated in favor of moment#subtract(Number, String).

**Comment by duongdev** [comment]
@vhlam1997 please re-request me to review when you are ready

**Comment by lamvh** [comment]
Please re-review, I fixed your suggests.

File: package.json [hunk 11]
```diff
@@ -11,17 +11,18 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
-    "@types/dotenv": "^6.1.1",
     "@types/express": "^4.16.1",
```
Comment (duongdev) [inline package.json]: all the @types/* should be on devDependencies, not only @types/dotenv

File: src/utils/xlsx.ts [hunk 12]
```diff
@@ -1,11 +1,23 @@
...
-}
\ No newline at end of file
+import * as xlsx from 'node-xlsx'
+import fs from 'fs'
+
+if (process.env.NODE_ENV === 'development') {
+  require('dotenv').config()
+}
+
+const DIR_NAME = process.env.DIR_NAME || 'exports'
```
Comment (duongdev) [inline src/utils/xlsx.ts]: still `DIR_NAME` here?

File: .gitignore [hunk 13]
```diff
@@ -1,3 +1,7 @@
 node_modules/
 dist/
 .env
+docker-compose.yml
```
Comment (duongdev) [inline .gitignore:4]: why do you ignore these files/dirs?

https://github.com/targeek/slm-data-exporter/blob/4a147a1fa5df8c52873e7fb00e0801ae0530b497/.gitignore#L4-L7

File: package.json (hunk 11 above)
Comment (duongdev) [reply package.json]: https://github.com/targeek/slm-data-exporter/blob/4a147a1fa5df8c52873e7fb00e0801ae0530b497/package.json#L14-L15

**Review by duongdev — CHANGES_REQUESTED** [review]

**Review by duongdev — APPROVED** [review]

**Pull request merged**